internal/wasi/fs.go

```diff
@@ -111,17 +148,159 @@ func (f *memFile) Seek(offset int64, whence int) (int64, error) {
+	// The result of ReadDir is undefined in POSIX in that situation, so WASI will be the same.
+	entries := make([]fs.DirEntry, 0)
+	for name, entry := range dir.Entries {
+		entries = append(entries, &memFileInfo{name: path.Base(name), fsEntry: entry})
+	}
+	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
+
+	return &memDir{
+		name:    baseName,
+		entries: entries,
+		fsEntry: dir,
 	}
+}
+
+// Stat implements fs.File.Stat
+func (d *memDir) Stat() (fs.FileInfo, error) {
+	return &memFileInfo{name: d.name, fsEntry: d.fsEntry}, nil
+}
+
+// Read implements fs.File.Read
+func (d *memDir) Read(p []byte) (int, error) {
+	return 0, fmt.Errorf("read is not supported by directory")
+}
+
+// Close implements fs.File.Close
+func (d *memDir) Close() error {
+	return nil
+}
+
+// Seek implements io.Seeker
+func (d *memDir) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		d.offset = offset
+	case io.SeekCurrent:
+		d.offset += offset
+	case io.SeekEnd:
+		d.offset = int64(len(d.entries)) + offset
+	default:
+		return 0, fmt.Errorf("invalid whence: %d", whence)
+	}
+	return d.offset, nil
+}
+
+// ReadDir implements fs.ReadDirFile.
+func (d *memDir) ReadDir(n int) ([]fs.DirEntry, error) {
+	remaining := int64(len(d.entries)) - d.offset
+	if remaining == 0 &&
+		n > 0 /* only when n > 0, since ReadDir should return empty slice instead of io.EOF when n <= 0 */ {
+		return []fs.DirEntry{}, io.EOF
+	}
+
+	if n <= 0 || n > int(remaining) {
+		n = int(remaining)
+	}
+	entries := d.entries[d.offset : d.offset+int64(n)]
+	d.offset += int64(n)
+
+	return entries, nil
+}
+
+// memFileInfo represents a FileInfo of an opened memFile or memDir.
+// memFileInfo implements fs.FileInfo and fs.DirEntry
+type memFileInfo struct {
+	// name of this file, not a full path but a base name.
+	name string
+	// The memFSEntry this memFileInfo is about.
+	fsEntry *memFSEntry
+}
+
+// Name implements fs.FileInfo.Name
+func (e *memFileInfo) Name() string { return e.name }
+
+// Size implements fs.FileInfo.Size
+func (e *memFileInfo) Size() int64 { return int64(len(e.fsEntry.Contents)) }
+
+// Mode implements fs.FileInfo.Mode
+func (e *memFileInfo) Mode() fs.FileMode {
+	mode := fs.FileMode(0777)
+	if e.fsEntry.IsDir() {
+		mode |= fs.ModeDir
+	}
+	return mode
+}
+
+// Type implements fs.FileInfo.Type
+func (e *memFileInfo) Type() fs.FileMode { return e.Mode().Type() }
+
+// ModTime implements fs.FileInfo.ModTime
+func (e *memFileInfo) ModTime() time.Time { return time.Time{} } // return the empty value for now
+
+// IsDir implements fs.FileInfo.IsDir
+func (f *memFileInfo) IsDir() bool { return f.Mode().IsDir() }
+
+// Sys implements fs.FileInfo.Sys
+func (f *memFileInfo) Sys() interface{} { return nil }
+
+// Info implements fs.DirEntry.Info
+func (f *memFileInfo) Info() (fs.FileInfo, error) { return f, nil }
+
+// readerWriterFile wraps io.Reader and io.Writer as fs.File.
```

Comment:
I don't really understand the relevance to "as fs.File" because fs.File has a different signature. It is also confusing that reader and writer are exported.

If you want to make a utility that is readerWriterFile, maybe that comment would be better

```
// compile-time check that readerWriterFile wraps io.Reader and io.Writer as fs.File.
_ io.File = &readerWriterFile{}
```